@dotlambda Thank you for the review! I've fixed and updated.

> Please add pytestCheckHook and versionCheckHook.

If added `pytestCheckHook` cause build error.

<details>
<summary>build log (if added pythonCheckHook)</summary>

```
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing pypa-build-hook
Using pypaBuildPhase
Sourcing python-runtime-deps-check-hook
Using pythonRuntimeDepsCheckHook
Sourcing pypa-install-hook
Using pypaInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing pytest-check-hook
Using pytestCheckPhase
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/8drzxbjdj233cc93fwhgadyqiwi2risx-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file "source/uv.lock"
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
no configure script, doing nothing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Executing pypaBuildPhase
Creating a wheel...
pypa build flags: --no-isolation --outdir dist/ --wheel
* Getting build dependencies for wheel...
* Building wheel...
Successfully built babeldoc-0.5.21-py3-none-any.whl
Finished creating a wheel...
Finished executing pypaBuildPhase
Running phase: pythonRuntimeDepsCheckHook
@nix { "action": "setPhase", "phase": "pythonRuntimeDepsCheckHook" }
Executing pythonRuntimeDepsCheck
Checking runtime dependencies for babeldoc-0.5.21-py3-none-any.whl
Finished executing pythonRuntimeDepsCheck
Running phase: installPhase
@nix { "action": "setPhase", "phase": "installPhase" }
Executing pypaInstallPhase
Successfully installed babeldoc-0.5.21-py3-none-any.whl
Finished executing pypaInstallPhase
Running phase: pythonOutputDistPhase
@nix { "action": "setPhase", "phase": "pythonOutputDistPhase" }
Executing pythonOutputDistPhase
Finished executing pythonOutputDistPhase
Running phase: fixupPhase
@nix { "action": "setPhase", "phase": "fixupPhase" }
checking for references to /nix/var/nix/builds/nix-11634-1144407885/ in /nix/store/4113nxg1ayl3j89brg6fwfp8p4x5hdib-babeldoc-0.5.21...
patching script interpreter paths in /nix/store/4113nxg1ayl3j89brg6fwfp8p4x5hdib-babeldoc-0.5.21
/nix/store/4113nxg1ayl3j89brg6fwfp8p4x5hdib-babeldoc-0.5.21/lib/python3.13/site-packages/babeldoc/pdfminer/psparser.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/bin/python3"
stripping (with command strip and flags -S) in  /nix/store/4113nxg1ayl3j89brg6fwfp8p4x5hdib-babeldoc-0.5.21/lib /nix/store/4113nxg1ayl3j89brg6fwfp8p4x5hdib-babeldoc-0.5.21/bin
checking for references to /nix/var/nix/builds/nix-11634-1144407885/ in /nix/store/kns8d4f4n2sinaq0z8ih3vc63nckypza-babeldoc-0.5.21-dist...
patching script interpreter paths in /nix/store/kns8d4f4n2sinaq0z8ih3vc63nckypza-babeldoc-0.5.21-dist
Rewriting #!/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/bin/python3.13 to #!/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9
wrapping `/nix/store/4113nxg1ayl3j89brg6fwfp8p4x5hdib-babeldoc-0.5.21/bin/babeldoc'...
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
Running phase: installCheckPhase
@nix { "action": "setPhase", "phase": "installCheckPhase" }
no Makefile or custom installCheckPhase, doing nothing
Running phase: pythonCatchConflictsPhase
@nix { "action": "setPhase", "phase": "pythonCatchConflictsPhase" }
Running phase: pythonRemoveBinBytecodePhase
@nix { "action": "setPhase", "phase": "pythonRemoveBinBytecodePhase" }
Running phase: pythonImportsCheckPhase
@nix { "action": "setPhase", "phase": "pythonImportsCheckPhase" }
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: babeldoc
Running phase: pytestCheckPhase
@nix { "action": "setPhase", "phase": "pytestCheckPhase" }
Executing pytestCheckPhase
pytest flags: -m pytest
============================= test session starts ==============================
platform darwin -- Python 3.13.9, pytest-8.4.2, pluggy-1.6.0
rootdir: /nix/var/nix/builds/nix-11634-1144407885/source
configfile: pyproject.toml
testpaths: tests
plugins: anyio-4.11.0
collecting ... 
collected 0 items / 1 error                                                    

==================================== ERRORS ====================================
___________ ERROR collecting tests/test_translation_cache_cleanup.py ___________
/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/lib/python3.13/pathlib/_local.py:722: in mkdir
    os.mkdir(self, mode)
E   FileNotFoundError: [Errno 2] No such file or directory: '/homeless-shelter/.cache/babeldoc/tiktoken'

During handling of the above exception, another exception occurred:
/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/lib/python3.13/pathlib/_local.py:722: in mkdir
    os.mkdir(self, mode)
E   FileNotFoundError: [Errno 2] No such file or directory: '/homeless-shelter/.cache/babeldoc'

During handling of the above exception, another exception occurred:
/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/lib/python3.13/pathlib/_local.py:722: in mkdir
    os.mkdir(self, mode)
E   FileNotFoundError: [Errno 2] No such file or directory: '/homeless-shelter/.cache'

During handling of the above exception, another exception occurred:
tests/test_translation_cache_cleanup.py:3: in <module>
    from babeldoc.translator.cache import TranslationCache
babeldoc/translator/cache.py:16: in <module>
    from babeldoc.const import CACHE_FOLDER
babeldoc/const.py:43: in <module>
    TIKTOKEN_CACHE_FOLDER.mkdir(parents=True, exist_ok=True)
/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/lib/python3.13/pathlib/_local.py:726: in mkdir
    self.parent.mkdir(parents=True, exist_ok=True)
/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/lib/python3.13/pathlib/_local.py:726: in mkdir
    self.parent.mkdir(parents=True, exist_ok=True)
/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/lib/python3.13/pathlib/_local.py:726: in mkdir
    self.parent.mkdir(parents=True, exist_ok=True)
/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/lib/python3.13/pathlib/_local.py:722: in mkdir
    os.mkdir(self, mode)
E   OSError: [Errno 30] Read-only file system: '/homeless-shelter'
=========================== short test summary info ============================
ERROR tests/test_translation_cache_cleanup.py - OSError: [Errno 30] Read-only file system: '/homeless-shelter'
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.29s ===============================
```

</details>


<details>
<summary>build log (if added versionCheckHook)</summary>

```
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing pypa-build-hook
Using pypaBuildPhase
Sourcing python-runtime-deps-check-hook
Using pythonRuntimeDepsCheckHook
Sourcing pypa-install-hook
Using pypaInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
Using versionCheckHook
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/8drzxbjdj233cc93fwhgadyqiwi2risx-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file "source/uv.lock"
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
no configure script, doing nothing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Executing pypaBuildPhase
Creating a wheel...
pypa build flags: --no-isolation --outdir dist/ --wheel
[1m* Getting build dependencies for wheel...[0m
[1m* Building wheel...[0m
[1m[92mSuccessfully built [4mbabeldoc-0.5.21-py3-none-any.whl[0m[1m[92m[0m
Finished creating a wheel...
Finished executing pypaBuildPhase
Running phase: pythonRuntimeDepsCheckHook
@nix { "action": "setPhase", "phase": "pythonRuntimeDepsCheckHook" }
Executing pythonRuntimeDepsCheck
Checking runtime dependencies for babeldoc-0.5.21-py3-none-any.whl
Finished executing pythonRuntimeDepsCheck
Running phase: installPhase
@nix { "action": "setPhase", "phase": "installPhase" }
Executing pypaInstallPhase
Successfully installed babeldoc-0.5.21-py3-none-any.whl
Finished executing pypaInstallPhase
Running phase: pythonOutputDistPhase
@nix { "action": "setPhase", "phase": "pythonOutputDistPhase" }
Executing pythonOutputDistPhase
Finished executing pythonOutputDistPhase
Running phase: fixupPhase
@nix { "action": "setPhase", "phase": "fixupPhase" }
checking for references to /nix/var/nix/builds/nix-81578-1604113615/ in /nix/store/5nvz27njvc4m4h23j8zr3r4p26j7w5z7-babeldoc-0.5.21...
patching script interpreter paths in /nix/store/5nvz27njvc4m4h23j8zr3r4p26j7w5z7-babeldoc-0.5.21
/nix/store/5nvz27njvc4m4h23j8zr3r4p26j7w5z7-babeldoc-0.5.21/lib/python3.13/site-packages/babeldoc/pdfminer/psparser.py: interpreter directive changed from "#!/usr/bin/env python3" to "/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/bin/python3"
stripping (with command strip and flags -S) in  /nix/store/5nvz27njvc4m4h23j8zr3r4p26j7w5z7-babeldoc-0.5.21/lib /nix/store/5nvz27njvc4m4h23j8zr3r4p26j7w5z7-babeldoc-0.5.21/bin
checking for references to /nix/var/nix/builds/nix-81578-1604113615/ in /nix/store/45mdmy5igscpfy6gpqhn5rz3jvvsr7kg-babeldoc-0.5.21-dist...
patching script interpreter paths in /nix/store/45mdmy5igscpfy6gpqhn5rz3jvvsr7kg-babeldoc-0.5.21-dist
Rewriting #!/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9/bin/python3.13 to #!/nix/store/84s01mzl74vh69aphpl3rf07pp49pkz0-python3-3.13.9
wrapping `/nix/store/5nvz27njvc4m4h23j8zr3r4p26j7w5z7-babeldoc-0.5.21/bin/babeldoc'...
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
Running phase: installCheckPhase
@nix { "action": "setPhase", "phase": "installCheckPhase" }
Executing versionCheckPhase
Did not find version 0.5.21 in the output of the command /nix/store/5nvz27njvc4m4h23j8zr3r4p26j7w5z7-babeldoc-0.5.21/bin/babeldoc --version
Traceback (most recent call last):
  File "{{storeDir}}/lib/python3.13/pathlib/_local.py", line 722, in mkdir
    os.mkdir(self, mode)
    ~~~~~~~~^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/var/empty/.cache/babeldoc/tiktoken'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "{{storeDir}}/lib/python3.13/pathlib/_local.py", line 722, in mkdir
    os.mkdir(self, mode)
    ~~~~~~~~^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/var/empty/.cache/babeldoc'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "{{storeDir}}/bin/.babeldoc-wrapped", line 6, in <module>
    from babeldoc.main import cli
  File "{{storeDir}}/lib/python3.13/site-packages/babeldoc/main.py", line 19, in <module>
    import babeldoc.assets.assets
  File "{{storeDir}}/lib/python3.13/site-packages/babeldoc/assets/assets.py", line 25, in <module>
    from babeldoc.const import get_cache_file_path
  File "{{storeDir}}/lib/python3.13/site-packages/babeldoc/const.py", line 43, in <module>
    TIKTOKEN_CACHE_FOLDER.mkdir(parents=True, exist_ok=True)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "{{storeDir}}/lib/python3.13/pathlib/_local.py", line 726, in mkdir
    self.parent.mkdir(parents=True, exist_ok=True)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "{{storeDir}}/lib/python3.13/pathlib/_local.py", line 726, in mkdir
    self.parent.mkdir(parents=True, exist_ok=True)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "{{storeDir}}/lib/python3.13/pathlib/_local.py", line 722, in mkdir
    os.mkdir(self, mode)
    ~~~~~~~~^^^^^^^^^^^^
PermissionError: [Errno 1] Operation not permitted: '/var/empty/.cache'
```

</details>

--- a/pkgs/by-name/ba/babeldoc/package.nix
+++ b/pkgs/by-name/ba/babeldoc/package.nix
@@ -1,0 +1,90 @@
+{
+  lib,
+  python3Packages,
+  fetchFromGitHub,
+  fetchpatch2,
+  writableTmpDirAsHomeHook,
+  versionCheckHook,
+}:
+
+python3Packages.buildPythonApplication rec {
+  pname = "babeldoc";
+  version = "0.5.22";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "funstory-ai";
+    repo = "BabelDOC";
+    tag = "v${version}";
+    hash = "sha256-ArLTv5AjpUdbsN8bQs03ATwg5ugXetld2FmHhicU8OE=";
+  };
+
+  patches = [
+    (fetchpatch2 {
+      name = "rename-python-levenshtein-to-levenshtein";
+      url = "https://github.com/funstory-ai/BabelDOC/pull/542.patch?full_index=1";
+      hash = "sha256-rjXhKVFivkJo54WdYiihqB3lrlu4YEwVZZkE4WBatWs=";
+    })
+  ];
+
+  build-system = with python3Packages; [ hatchling ];
+
+  dependencies =
+    with python3Packages;
+    [
+      bitstring
+      configargparse
+      httpx
+      huggingface-hub
+      numpy
+      onnx
+      onnxruntime
+      openai
+      orjson
+      charset-normalizer
+      cryptography
+      peewee
+      psutil
+      pymupdf
+      rich
+      toml
+      tqdm
+      xsdata
+      msgpack
+      pydantic
+      tenacity
+      scikit-image
+      freetype-py
+      tiktoken
+      levenshtein
+      opencv-python-headless
+      rapidocr-onnxruntime
+      pyzstd
+      hyperscan
+      rtree
+      chardet
+      scipy
+      uharfbuzz
+      scikit-learn
+    ]
+    ++ httpx.optional-dependencies.socks
+    ++ (with xsdata.optional-dependencies; cli ++ lxml ++ soap);
+
+  pythonImportsCheck = [ "babeldoc" ];
+
+  nativeCheckInputs = [
+    writableTmpDirAsHomeHook
+    python3Packages.pytestCheckHook
+    versionCheckHook
+  ];
+  versionCheckKeepEnvironment = "HOME";
+
+  meta = {
+    description = "PDF scientific paper translation and bilingual comparison library";
+    homepage = "https://github.com/funstory-ai/BabelDOC";
+    changelog = "https://github.com/funstory-ai/BabelDOC/releases/tag/${src.tag}";
+    license = lib.licenses.agpl3Only;
+    maintainers = with lib.maintainers; [ ryota2357 ];
+    mainProgram = "babeldoc";
+  };
+}

